### ISSUE TITLE
ScrollView weird behaviour when there are less items

### ISSUE THREAD

**Issue by shreyanshgoel9** [facebook/react-native#42874]
### Description

When the scrollview container is bigger than total item width, on scroll the items start sticking on right on android.
Tested on physical device as well

https://github.com/facebook/react-native/assets/136585177/6ed7efe5-9436-46c5-b5d8-e34236b496af

### Steps to reproduce

1. Installed React Native macOs, android simulator
2. Write this code
` <ScrollView horizontal={true} style={{backgroundColor: 'green'}}>
        <View
          style={{
            height: 40,
            width: 50,
            marginRight: 20,
            backgroundColor: 'red',
          }}
        />
        <View
          style={{
            height: 40,
            width: 50,
            marginRight: 20,
            backgroundColor: 'red',
          }}
        />
      </ScrollView>`
      
      
3. Scroll horizontal to reproduce

### React Native Version

0.73.3

### Affected Platforms

Runtime - Android

### Output of `npx react-native info`

```text
System:
  OS: macOS 14.1.2
  CPU: (8) arm64 Apple M1 Pro
  Memory: 87.98 MB / 16.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 18.12.1
    path: ~/.nvm/versions/node/v18.12.1/bin/node
  Yarn:
    version: 1.22.19
    path: ~/.nvm/versions/node/v18.12.1/bin/yarn
  npm:
    version: 8.19.2
    path: ~/.nvm/versions/node/v18.12.1/bin/npm
  Watchman:
    version: 2023.08.14.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.14.2
    path: /opt/homebrew/opt/ruby/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.2
      - iOS 17.2
      - macOS 14.2
      - tvOS 17.2
      - visionOS 1.0
      - watchOS 10.2
  Android SDK: Not Found
IDEs:
  Android Studio: 2022.3 AI-223.8836.35.2231.10811636
  Xcode:
    version: 15.2/15C500b
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 20.0.2
    path: /opt/homebrew/opt/openjdk/bin/javac
  Ruby:
    version: 3.2.2
    path: /opt/homebrew/opt/ruby/bin/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.73.3
    wanted: 0.73.3
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: false
```

### Stacktrace or Logs

```text
-
```

### Reproducer

[snack.expo.dev/@uvrubel/rn-scrollview-horizontal](https://snack.expo.dev/@uvrubel/rn-scrollview-horizontal)

**Comment by shreyanshgoel9** [comment]
Snack
https://snack.expo.dev/@shreyanshgoel9/8cd020?platform=android

**Comment by Myst1024** [comment]
Running into the same issue, using React Native 0.73.4

**Comment by RakaDoank** [comment]
I also have this issue. I can reproduce this in physical Android device (API 33) and Android simulator (API 34). So far, no problem on iOS simulator.

I actually wanted to write this earlier. I remember i get this issue since i upgraded the RN to 0.73.0 from 0.72.7, until now currently on 0.73.4.

**Comment by kemplaw** [comment]
Running into the same issue, using React Native 0.72.10.

**Comment by ulyavrubel** [comment]
Running into the same issue, using React Native 0.73.4. Reproducible on Android only

## Steps to Reproduce
```
     <ScrollView horizontal>
       ...
      </ScrollView>
```
      
## Demo
https://snack.expo.dev/@uvrubel/rn-scrollview-horizontal


https://github.com/facebook/react-native/assets/52666724/1f17dabd-1832-4089-8787-e38b028b7410

**Comment by Titozzz** [comment]
Can confirm this is not something that was here before (but cannot easily bisect).
I can reproduce on 0.73.2 and 0.73.6

**Comment by Titozzz** [comment]
https://github.com/facebook/react-native/commit/30c7e9dfa41348e96e946384c76f038ccd859896 is the culprit!

**Comment by NickGerleman** [comment]
Whoah, that's a crazy effect.

The interesting logic here is gated on:

`mReactScrollViewScrollState.getLayoutDirection() == LAYOUT_DIRECTION_RTL`
https://github.com/facebook/react-native/blob/30c7e9dfa41348e96e946384c76f038ccd859896/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactHorizontalScrollView.java#L1308

@Titozzz is this path being hit on physical device?

**Comment by Titozzz** [comment]
I'll have to check tomorrow ! I'm home now 😬

**Comment by Titozzz** [comment]
https://github.com/facebook/react-native/commit/30c7e9dfa41348e96e946384c76f038ccd859896#diff-067e69cd88e616e9605533a4be612685b607a7d1cd29f854ee77d713113cb170R612-R614 is the change that breaks things

**Comment by NickGerleman** [comment]
Huh. That was to fix another bug, where flinging to the edge would sometimes start scrolling backwards (https://github.com/facebook/react-native/issues/38236 is one report, though I think I saw others after the fact).

The version clamp was based on a linked Android bug report being fixed, and seeing incorrect behavior in emulator. So I think there might be a couple possibilities:
1. Different behavior on another Android version. @Titozzz do you remember version in emulator vs device where you could or couldn’t repro this? I we can only repro on some devices, I wonder if this might be the case. 
2. There are more bugs here, that the issue ended up masking.

**Comment by Regin99** [comment]
> [30c7e9d#diff-067e69cd88e616e9605533a4be612685b607a7d1cd29f854ee77d713113cb170R612-R614](https://github.com/facebook/react-native/commit/30c7e9dfa41348e96e946384c76f038ccd859896#diff-067e69cd88e616e9605533a4be612685b607a7d1cd29f854ee77d713113cb170R612-R614) is the change that breaks things

Even if i fully downgrade android scroll logic (ReactAndroid/src/.../views/scroll) to 0.72.4 version, bug is still reproduced. Interesting fact, that shopify's flashlist reproduce that bug too. 
Are u sure that this changes breaks things?

**Comment by Titozzz** [comment]
Yep you are using prefab probably! 
https://reactnative.dev/contributing/how-to-build-from-source#update-your-project-to-build-from-source
Did you do that ?

**Comment by Titozzz** [comment]
This is the line that breaks it for me https://github.com/facebook/react-native/commit/30c7e9dfa41348e96e946384c76f038ccd859896#diff-067e69cd88e616e9605533a4be612685b607a7d1cd29f854ee77d713113cb170R612-R614. 

(Verified with breakpoints)

**Comment by Regin99** [comment]
> Yep you are using prefab probably! https://reactnative.dev/contributing/how-to-build-from-source#update-your-project-to-build-from-source Did you do that ?

Yeah, my mistake. I'm forgot bout prefabs. Thank you

**Comment by hjun555** [comment]
same issue here, any solutions?

**Comment by Regin99** [comment]
> same issue here, any solutions?

Look above

**Comment by baez97** [comment]
In case someone needs a fix without upgrading, setting `contentContainerStyle` to full width worked for me
```jsx
<ScrollView horizontal contentContainerStyle={{ width: "100%" }}>
```

**Comment by RobinBobin** [comment]
Was it fixed in both 0.73 and 0.74? What are the versions, please?
I still see this bug in 0.73.4.

**Comment by NickGerleman** [comment]
> Was it fixed in both 0.73 and 0.74? What are the versions, please? I still see this bug in 0.73.4.

Available in 0.73.7+

**Comment by Arjun059** [comment]
Quick Fix;
  <ScrollView
    horizontal={true}
    contentContainerStyle={{flexGrow: 1}}
  >
         {/*  Your content */}
  </ScrollView >
